extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java

```diff
@@ -222,32 +223,68 @@ Most of the window operations like SUM(), RANK(), RANGE() etc. can be made with
         final Frame frame = inputChannel.read();
         convertRowFrameToRowsAndColumns(frame);
         return ReturnOrAwait.runAgain();
-      } else if (inputChannel.isFinished()) {
-        runAllOpsOnMultipleRac(frameRowsAndCols);
+      }
+
+      if (inputChannel.isFinished()) {
+        // If there are rows pending flush, flush them and run again.
+        if (frameHasRowsPendingFlush()) {
+          return flushRACsAndRunAgain();
+        }
+
+        // If no rows are flushed yet, process all rows.
+        if (rowId.get() == 0) {
+          runAllOpsOnMultipleRac(frameRowsAndCols);
+        }
+
+        // If there are still rows pending after operations, run again.
+        if (frameHasRowsPendingFlush()) {
+          return ReturnOrAwait.runAgain();
+        }
         return ReturnOrAwait.returnObject(Unit.instance());
-      } else {
-        return ReturnOrAwait.awaitAll(inputChannels().size());
       }
+      return ReturnOrAwait.awaitAll(inputChannels().size());
     }
 
     // Scenario 2: All window functions in the query have OVER() clause with a PARTITION BY
     if (frameCursor == null || frameCursor.isDone()) {
       if (readableInputs.isEmpty()) {
         return ReturnOrAwait.awaitAll(1);
-      } else if (inputChannel.canRead()) {
+      }
+
+      if (inputChannel.canRead()) {
         final Frame frame = inputChannel.read();
         frameCursor = FrameProcessors.makeCursor(frame, frameReader);
         makeRowSupplierFromFrameCursor();
       } else if (inputChannel.isFinished()) {
-        // Handle any remaining data.
-        lastPartitionIndex = rowsToProcess.size() - 1;
-        processRowsUpToLastPartition();
+        // If we have some rows pending processing, process them.
+        // We run it again as it's possible that frame writer's capacity got reached and some output rows are
+        // pending flush to the output channel.
+        if (!rowsToProcess.isEmpty()) {
+          lastPartitionIndex = rowsToProcess.size() - 1;
+          processRowsUpToLastPartition();
+          return ReturnOrAwait.runAgain();
+        }
+
+        // This handles the rows pending flush to the output channel, if we have any.
+        if (frameHasRowsPendingFlush()) {
+          return flushRACsAndRunAgain();
+        } else {
+          clearRACBuffers();
+        }
```

Comment:
why do we have to come in here 2 times to do a `flush`?

why not change `processRowsUpToLastPartition()`  to have a flush argument; so 
`processRowsUpToLastPartition(true)`  would flush and cleanup?

Reply:
> why do we have to come in here 2 times to do a flush?

Because it's possible that not all the rows have been flushed to the output channel, if we hit the frame writer's capacity.

> why not change processRowsUpToLastPartition() to have a flush argument; so
processRowsUpToLastPartition(true) would flush and cleanup?

We don't just need to flush and cleanup, we also need to check the frame writer if we want to re-run `runIncrementally()` again --- as flush doesn't ensure that all rows have been flushed to the output channel.

For example, if frame writer is allowing 20k rows as the capacity, and the input frame has 80k rows, then we need 4 iterations of `runIncrementally()`.